File: 5-Problem-A-Day/114. Flatten Binary Tree to Linked List.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public void flatten(TreeNode root) {
        if (root == null) return;
        Stack<TreeNode> nodes = new Stack<>();
        nodes.push(root);
        TreeNode prev = null;
        while (!nodes.empty()) {
            TreeNode node = nodes.pop();
            if (prev == null)
                prev = node;
            else {
                prev.right = node;
                prev = prev.right;
            }
            while (node.left != null) {
                if (node.right != null)
                    nodes.push(node.right);                
                TreeNode tmp = node;
                node = node.left;   
                tmp.left = null;
                prev.right = node;
                prev = prev.right;
            }
            if (node.right != null)
                nodes.push(node.right); 
        }        
    }
}

File: 2023/114. Flatten Binary Tree to Linked List.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void flatten(TreeNode* root) {
        TreeNode* prev = NULL;
        preorder(root, prev);
    }

    void preorder(TreeNode* root, TreeNode*& prev) {
        if (root == NULL) return;
        TreeNode* left = root->left;
        TreeNode* right = root->right;
        if (prev != NULL) {
            prev->right = root;
        }
        prev = root;
        root->left = root->right = NULL;
        preorder(left, prev);
        preorder(right, prev);
    }
};